### PR TITLE
MISUV-7486: Select Your Reason Page

### PR DESCRIPTION
**Pull request by owen-gilfellon**
- Added Select Your Reason page
- Added routes, and routes for CYA
- Brought in some classes from HMRC scaffold and associated tests
- Renamed claimToAdjustPOA package to claimToAdjustPoa for consistency, and seemed to fix issues running tests via IntelliJ
-

### CODE REVIEW

File: test/controllers/claimToAdjustPoa/SelectYourReasonControllerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            .withSession("nino" -> BaseTestConstants.testNino, "origin" -> "PTA")
+
+          val result = TestSelectYourReasonController.submit(isAgent = false, isChange = false)(request)
+
+          status(result) shouldBe SEE_OTHER
+          redirectLocation(result) shouldBe Some("/report-quarterly/income-and-expenses/view/adjust-poa/check-your-answers")
+        }
+      }
+
+      "in check mode" when {
```
Comment (avelx) [inline test/controllers/claimToAdjustPoa/SelectYourReasonControllerSpec.scala:220]: I would assume this UTs is still something in TODO?

Comment (owen-gilfellon) [reply test/controllers/claimToAdjustPoa/SelectYourReasonControllerSpec.scala:220]: I think there's a ticket for doing CYA / change answers.

File: app/controllers/claimToAdjustPoa/SelectYourReasonController.scala [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    val errorHandler = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+
+    for {
+      session <- EitherT(sessionService.getMongo)
+      poa <- EitherT(claimToAdjustService.getPoaForNonCrystallisedTaxYear(Nino(user.nino)))
+      result <- (session, poa) match {
+        case (Some(s), Some(p)) =>
+          block(s, p)
+        case (None, _) =>
+          Logger("application").error(s"[SelectYourReasonController][withSessionAndPoa] session missing")
```
Comment (avelx) [inline app/controllers/claimToAdjustPoa/SelectYourReasonController.scala:109]: no need to specify source of the problem in the log entry, I think we recently merge this change and it manager by logback configuration, so you can drop prefix like: "[SelectYourReasonController][withSessionAndPoa].."

File: test/controllers/claimToAdjustPoa/SelectYourReasonControllerSpec.scala (hunk 1 above)
Comment (avelx) [reply test/controllers/claimToAdjustPoa/SelectYourReasonControllerSpec.scala:220]: If we can refer to specific story in these TODO, as its a bit confusing at the moment and we can easily forgot about these in the future; there are quite a few TODO on line 104 and 108 in this file

File: app/forms/mappings/Constraints.scala [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+        import ev._
+
+        if (input <= maximum) {
+          Valid
+        } else {
+          Invalid(errorKey, maximum)
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (owen-gilfellon) [reply app/forms/mappings/Constraints.scala:59]: They come as standard with the scaffold - my preference would be to leave them uncommented, and perhaps do a tech talk about the scaffold and how to use the new classes.

File: test/generators/Generators.scala [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  def stringsLongerThan(minLength: Int): Gen[String] = for {
+    maxLength <- (minLength * 2).max(100)
+    length    <- Gen.chooseNum(minLength + 1, maxLength)
+    chars     <- listOfN(length, arbitrary[Char])
+  } yield chars.mkString
+
+  def stringsExceptSpecificValues(excluded: Seq[String]): Gen[String] =
+    nonEmptyString suchThat (!excluded.contains(_))
+
+  def oneOf[T](xs: Seq[Gen[T]]): Gen[T] =
```
Comment (avelx) [inline test/generators/Generators.scala:102]: I think this is something already provided by the ScalaCheck: have a look at the usages of Gen.oneOf

File: test/forms/mappings/ConstraintsSpec.scala [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      val result = maxLength(10, "error.length")("a" * 11)
+      result mustEqual Invalid("error.length", 10)
+    }
+  }
+
+  "maxDate" - {
+
+    "must return Valid for a date before or equal to the maximum" in {
+
+      val gen: Gen[(LocalDate, LocalDate)] = for {
```
Comment (avelx) [inline test/forms/mappings/ConstraintsSpec.scala:130]: this date generator seems to be duplicated across this spec, would be nice to move it under main generator

File: app/forms/mappings/Constraints.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (avelx) [reply app/forms/mappings/Constraints.scala:59]: yeh, demo or discussion during tech deep dive would be ideal, this approach is an improvement for sure, but yes, let have a discussion on this next week

**Review by avelx — APPROVED** [review]
approved with comments

**Pull request merged**